SonarQube.TeamBuild.PreProcessor/Roslyn/PluginResourceCache.cs

```diff
@@ -0,0 +1,118 @@
+﻿/*
+ * SonarQube Scanner for MSBuild
+ * Copyright (C) 2015-2017 SonarSource SA and Microsoft Corporation
+ * mailto: contact AT sonarsource DOT com
+ *
+ * Licensed under the MIT License.
+ * See LICENSE file in the project root for full license information.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+using SonarQube.Common;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Newtonsoft.Json;
+
+namespace SonarQube.TeamBuild.PreProcessor.Roslyn
+{
+    class SubdirIndex
+    {
+        private readonly string basedir;
+        private readonly string indexPath;
+        private readonly IDictionary<string, string> mapping;
+
+        public SubdirIndex(string basedir)
+        {
+            this.basedir = basedir;
+            this.indexPath = Path.Combine(basedir, "index.json");
+            this.mapping = Read();
+        }
+
+        public IDictionary<string, string> Read()
+        {
+            if (File.Exists(indexPath))
+            {
+                return JsonConvert.DeserializeObject<IDictionary<string, string>>(File.ReadAllText(indexPath));
+            }
+            return new Dictionary<string, string>();
+        }
+
+        private void Flush()
+        {
+            File.WriteAllText(indexPath, JsonConvert.SerializeObject(mapping));
+        }
+
+        public string GetOrCreatePath(string key)
+        {
+            lock (mapping)
```

Comment:
Not sure about the synchronize in Java so I am giving you a link to the explanation of the [`lock` in c#](https://msdn.microsoft.com/en-us/library/c5kehkcz.aspx) but I guess this is doing what you want. The only thing I would say is that I am not entirely sure if it is safe to lock on the `mapping` object, I have always seen people creating a special object like this `private Object thisLock = new Object();`.